File: Tree-shaped Deep Forest/Trie/Q1. Longest Common Prefix/Q1_LongestCommonPrefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        // Edge case: if array is empty, return ""
        if (strs == null || strs.length == 0) {
            return "";
        }

        // Assume first string is the prefix
        String prefix = strs[0];

        // Loop over the rest of the strings
        for (int i = 1; i < strs.length; i++) {
            // While current string does not start with prefix
            while (strs[i].indexOf(prefix) != 0) {
                // Reduce prefix by removing last character
                prefix = prefix.substring(0, prefix.length() - 1);
                
                // If prefix becomes empty, no common prefix
                if (prefix.isEmpty()) {
                    return "";
                }
            }
        }

        return prefix;
    }
}

File: Tree-shaped Deep Forest/Trie/Q1. Longest Common Prefix/Q1_LongestCommonPrefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        # Step 1: If the list is empty, return ""
        if not strs:
            return ""

        # Step 2: Start with the first string as prefix
        prefix = strs[0]

        # Step 3: Compare prefix with each word in the list
        for word in strs[1:]:
            # While current word does not start with prefix, shrink prefix
            while not word.startswith(prefix):
                prefix = prefix[:-1]  # Remove last character
                # If prefix becomes empty, no common prefix exists
                if not prefix:
                    return ""

        # Step 4: Return the final common prefix
        return prefix
